crates/evm/traces/src/debug/mod.rs

```diff
@@ -0,0 +1,349 @@
+fn parse_types(source: &str) -> (Option<Parameters<'_>>, Option<Parameters<'_>>) {
+    let inputs = source.find('(').and_then(|params_start| {
+        let params_end = params_start + source[params_start..].find(')')?;
+        Parameters::parse(&source[params_start..params_end + 1]).ok()
+    });
+    let outputs = source.find("returns").and_then(|returns_start| {
+        let return_params_start = returns_start + source[returns_start..].find('(')?;
+        let return_params_end = return_params_start + source[return_params_start..].find(')')?;
+        Parameters::parse(&source[return_params_start..return_params_end + 1]).ok()
+    });
+
+    (inputs, outputs)
+}
+
+/// Given [Parameters] and [CallTraceStep], tries to decode parameters by using stack and memory.
+fn try_decode_args_from_step(args: &Parameters<'_>, step: &CallTraceStep) -> Option<Vec<String>> {
+    let params = &args.params;
+
+    if params.is_empty() {
+        return Some(vec![]);
+    }
+
+    let types = params.iter().map(|p| p.resolve().ok().map(|t| (t, p.storage))).collect::<Vec<_>>();
+
+    let stack = step.stack.as_ref()?;
+
+    if stack.len() < types.len() {
+        return None;
+    }
+
+    let inputs = &stack[stack.len() - types.len()..];
+
+    let decoded = inputs
+        .iter()
+        .zip(types.iter())
+        .map(|(input, type_and_storage)| {
+            type_and_storage
+                .as_ref()
+                .and_then(|(type_, storage)| {
+                    match (type_, storage) {
+                        // HACK: alloy parser treats user-defined types as uint8: https://github.com/alloy-rs/core/pull/386
+                        //
+                        // filter out `uint8` params which are marked as storage or memory as this
+                        // is not possible in Solidity and means that type is user-defined
+                        (DynSolType::Uint(8), Some(Storage::Memory | Storage::Storage)) => None,
+                        (_, Some(Storage::Memory)) => {
+                            decode_from_memory(type_, step.memory.as_bytes(), input.to::<usize>())
+                        }
+                        // Read other types from stack
+                        _ => type_.abi_decode(&input.to_be_bytes::<32>()).ok(),
+                    }
+                })
+                .as_ref()
+                .map(format_token)
+                .unwrap_or_else(|| "<unknown>".to_string())
+        })
+        .collect();
+
+    Some(decoded)
+}
+
+/// Decodes given [DynSolType] from memory.
+fn decode_from_memory(ty: &DynSolType, memory: &[u8], location: usize) -> Option<DynSolValue> {
+    let first_word = memory.get(location..location + 32)?;
+
+    match ty {
+        // For `string` and `bytes` layout is a word with length followed by the data
+        DynSolType::String | DynSolType::Bytes => {
+            let length = U256::from_be_bytes::<32>(first_word.try_into().unwrap()).to::<usize>();
```

Comment:
will the usize cause overflow problem, usize is 64 bits (8 bytes), where the length can be 32 bytes long

Reply:
i am trying this new feature, seeing below error in my end

```
Message:  Uint conversion error: Overflow(256, 10488988339550576416, 18446744073709551615)
Location: crates/evm/traces/src/debug/mod.rs:305

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 14 frames hidden ⋮
  15: core::result::Result<T,E>::expect::h1e62bb9f5f4f9d0f
      at /rustc/d124af4d4a707212dfe7c081/library/core/src/result.rs:1034
  16: ruint::from::<impl ruint::Uint<_,_>>::to::h7b50fa5729cd9809
      at /Users/x/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ruint-1.12.3/src/from.rs:236
       234 │         T: Debug,
       235 │     {
       236 >         self.uint_try_to().expect("Uint conversion error")
       237 │     }
       238 │
  17: foundry_evm_traces::debug::decode_from_memory::hb99e9e33dfb84c71
      at /Users/x/foundry/crates/evm/traces/src/debug/mod.rs:305
       303 │         // For `string` and `bytes` layout is a word with length followed by the data
       304 │         DynSolType::String | DynSolType::Bytes => {
       305 >             let length = U256::from_be_bytes::<32>(first_word.try_into().unwrap()).to::<usize>();
       306 │             let data = memory.get(location + 32..location + 32 + length)?;
       307 │
  18: foundry_evm_traces::debug::try_decode_args_from_step::{{closure}}::{{closure}}::h0912b74d67673459
      at /Users/x/foundry/crates/evm/traces/src/debug/mod.rs:283
       281 │                         (DynSolType::Uint(8), Some(Storage::Memory | Storage::Storage)) => None,
```

Reply:
hi @ChinW thanks for trying this out and reporting!

I think it's safe to assume that memory wouldn't be expanded for more than usize, though this analysis is not perfect and we might expect to find the length while there's actually some other value, which I believe has happened in your case

any chance you could provide a repro for this? also, are you using via-ir here?